middleware.go

```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
 			request.Latency = fmt.Sprintf("%.9fs", m.Duration.Seconds())
 			request.ResponseSize = strconv.FormatInt(m.Written, 10)
 
-			fields := logrus.Fields{"httpRequest": request}
-
-			traceHeader := r.Header.Get("X-Cloud-Trace-Context")
-			if traceHeader != "" {
-				fields["trace"] = traceHeader
+			if o.filterHTTP(r) {
+				// log the result
+				ctxlogrus.Extract(ctx).Infof("served HTTP %v %v", r.Method, r.URL)
 			}
+		})
+	}
+}
+
+func UnaryLoggingInterceptor(logger *logrus.Logger, opts ...MiddlewareOption) grpc.UnaryServerInterceptor {
+	o := evaluateMiddlewareOptions(opts)
+	return loggingInterceptor{logger: logger, middlewareOptions: o}.intercept
+}
+
+func StreamLoggingInterceptor(logger *logrus.Logger, opts ...MiddlewareOption) grpc.StreamServerInterceptor {
+	o := evaluateMiddlewareOptions(opts)
+	return loggingInterceptor{logger: logger, middlewareOptions: o}.interceptStream
+}
+
+type loggingInterceptor struct {
+	logger *logrus.Logger
+	*middlewareOptions
+}
+
+type GRPCRequest struct {
+	Method    string `json:"method,omitempty"`
+	UserAgent string `json:"userAgent,omitempty"`
+	PeerAddr  string `json:"peer,omitempty"`
+	Deadline  string `json:"deadline,omitempty"`
+	Duration  string `json:"duration,omitempty"`
+}
+
+func (l loggingInterceptor) intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (resp interface{}, err error) {
+	startTime := time.Now()
+	ctx = ctxlogrus.ToContext(ctx, logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	resp, err = handler(ctx, req)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
```

Comment:
why not milliseconds?  I hope we don't really need values over 1000 very often

Reply:
Copied straight from the source material:
https://github.com/StevenACoffman/logrus-stackdriver-formatter/blob/5eec3c1a5678addc707c79ecf92fe0b8d0f47ccd/middleware.go#L30

Theoretically GCP's log viewer can take correctly formatted latency field. 9 sigfigs is probably in excess, but it's taken directly from Google's docs on LogEntry's HttpRequest latency field:
https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#HttpRequest.FIELDS.latency

Though the HttpRequestContext used by ErrorEvent to enrich Error Reporting dashboard has no need for latency, and I've not yet seen if the LogEntry can be enriched directly with the HttpRequest in its jsonPayload field, such that you can get this neat little widget on the log entry like we have in load balancer logs:
![image](https://user-images.githubusercontent.com/4668480/111786407-23acda00-8894-11eb-87e2-825f7e7f50a7.png)

I'd like to have those, but once we have tracing to correlate all our logs to a single request context, it might not matter so much.

tl;dr: lmk if you feel strongly about this I can reduce it to 5 sigfigs like load balancer seems to use.

Reply:
I tested real quick promoting the httpRequest to the parent Entry to see if it Just Works. It does:
![image](https://user-images.githubusercontent.com/4668480/111816291-0a675600-88b3-11eb-836e-f85ee90fbbae.png)

But it overwrites the logged message with the requestURL, so I'm going to make a change that can promote this up only on these last "served {HTTP|RPC} [...]" entries.

Reply:
![image](https://user-images.githubusercontent.com/4668480/111834639-4659e580-88ca-11eb-84ab-d470c6588ffe.png)

Made it work, even for gRPC (by pretending it's close enough to HTTP)

Reply:
For consistency with how many sigfigs load balancer outputs, I'd want at least 5. In the grand scheme of things, I don't even care unless it's more than 3s. But I think you can also use this same field as-formatted to filter greater than some millisecond value same as I do in LB, e.g.:
```
trace:*
(httpRequest.latency>="3000ms" OR
jsonPayload.latency>="3000ms")
```

finds any entry with a traceID taking longer than 3s